### PR TITLE
Fix ICE reconnect flashings

### PR DESCRIPTION
**Pull request by keianhzo**
This PR adds support for better ICE failures recovery. Now instead of rejoining in case of an ICE failure we:
- Restart ICE for browser that support it
- Recreate the transports and recreate producers/consumers for browser that don't support ICE `updateIceServers`

This PR adds a couple of methods in the Dialog side for closing a transport and getting the consumers.

This has to be merged at the same time as: https://github.com/mozilla/dialog/pull/12

### CODE REVIEW

**Review by netpro2k — COMMENTED** [review]
I am still reviewing the actual ICE related changes in naf-dialog-adapter but wanted to submit the review comments I already have. Mostly just some minor styling feedback on the events system being used.

Starting to look at naf-dialog-adapter more, at first glance it feels like (even prior to this PR) we are doing a lot of work polling the connection state and trying to reconnect things.. I wonder if we are actually hurting ourselves, since I was under the impression that the ICE agent in the browser should already be handling that for us... I want to read through more of the mediasoup-client library to see if its already doing any sort of reconnect logic that we might be getting in the way of with our own reconnect logic.

I also want to read more about TURN and our coturn implementation in particular to understand why our TURN sessions would be expiring. Surely there is some way this is supposed to be getting refreshed... If we fix that then I don't think we have to worry about special handling for Firefox since updateIceServers should never need to be called, meaning we never have to recreate the transports, which would greatly simplify this PR.

File: src/components/media-views.js [hunk 1]
```diff
@@ -736,6 +736,21 @@ AFRAME.registerComponent("media-video", {
       if (url.startsWith("hubs://")) {
         const streamClientId = url.substring(7).split("/")[1]; // /clients/<client id>/video is only URL for now
         const stream = await NAF.connection.adapter.getMediaStream(streamClientId, "video");
+        // We subscribe to video stream notifications for this peer to update the video element
+        // This could happen in case there is an ICE failure that requires a transport recreation.
+        NAF.connection.adapter.addEventListener(streamClientId, "video", {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/components/media-views.js:741]: Fixed

File: src/components/media-views.js [hunk 2]
```diff
@@ -736,6 +736,21 @@ AFRAME.registerComponent("media-video", {
...
+          onStreamUpdated: async (peerId, kind) => {
+            if (peerId === streamClientId && kind === "video") {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/components/media-views.js:743]: That's right, we can safely remove that.

File: src/naf-dialog-adapter.js [hunk 3]
```diff
@@ -146,6 +150,24 @@ export default class DialogAdapter {
     this._onOccupantMessage = messageListener;
   }
 
+  // Add an event listener for a peer and an event type
+  addEventListener(peerId, eventType, listener) {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:154]: No reason at all, just what I came up with at the time. Regarding `EventEmitter` I saw it but I couldn't see it being used anywhere else but I'm happy to change to a well known and tested solution. For the scene events I considered it but just prefer each class to be responsible for it's own events rather than throwing everything inside the scene unless it's a global event that's going to be emitted across the board.

File: src/components/avatar-audio-source.js [hunk 4]
```diff
@@ -88,9 +92,28 @@ AFRAME.registerComponent("avatar-audio-source", {
...
+    // We subscribe to audio stream notifications for this peer to update the audio source
+    // This could happen in case there is an ICE failure that requires a transport recreation.
+    getOwnerId(this.el).then(peerId => {
+      NAF.connection.adapter.addEventListener(peerId, "audio", this);
+    });
     this.createAudio();
   },
 
+  async onStreamUpdated(peerId, kind) {
+    if (kind === "audio") {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/components/avatar-audio-source.js:104]: Moved to EventEmitter so I have kept that.

File: src/naf-dialog-adapter.js [hunk 5]
```diff
@@ -146,6 +150,24 @@ export default class DialogAdapter {
...
+      this._listeners.set(peerId, {
+        [eventType]: listener
+      });
+    } else {
+      this._listeners.get(peerId)[eventType] = listener;
+    }
+  }
+
+  // Removes an event listener for a peer and an event type
+  removeEventListener(peerId, eventType) {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:165]: I've moved this to eventemitter.

File: src/naf-dialog-adapter.js [hunk 6]
```diff
@@ -202,36 +224,91 @@ export default class DialogAdapter {
    * https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setConfiguration
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:224]: That's what I understand and what I saw during testing. When using TURN we need to update the ICE servers in case the credentials have expired otherwise we are good with just restarting ICE in both sides. As yous y say, if we are not using proxy candidates we can just restart ICE without updating the servers. 

We can also check here if we are we are using a non proxy candidate to avoid an ICE server update as even when forcing TURN we still add a STUN server. I was assuming here that if TURN is forced is because the client is having ICE issues using STUN.

File: src/components/avatar-audio-source.js [hunk 7]
```diff
@@ -88,9 +92,28 @@ AFRAME.registerComponent("avatar-audio-source", {
...
+      // The audio stream for this peer has been updated
+      const stream = await NAF.connection.adapter.getMediaStream(peerId, "audio").catch(e => {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/components/avatar-audio-source.js:106]: I agree, replacing tracks seems to be a better option. Updating.

File: src/components/audio-feedback.js [hunk 8]
```diff
@@ -68,7 +68,7 @@ AFRAME.registerComponent("networked-audio-analyser", {
         this.levels = new Uint8Array(this.analyser.fftSize);
         event.detail.soundSource.connect(this.analyser);
       },
-      { once: true }
+      { once: false }
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/components/audio-feedback.js:71]: Replacing tracks avoid dealing with this so reverting.

**Comment by keianhzo** [comment]
@netpro2k I'm happy to pair anytime and go over this and see if we can simplify it.

File: src/naf-dialog-adapter.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (netpro2k) [reply src/naf-dialog-adapter.js:208]: Even when a candidate selected via STUN fails the STUN server itself should not need to change so I think if we can fix the credentials issue we should not ever need to change the ICE servers (except in the case of a dialog server transfer

**Comment by keianhzo** [comment]
@netpro2k I've updated this PR to stop handling the disconnected state.

File: src/components/avatar-audio-source.js [hunk 9]
```diff
@@ -88,9 +95,34 @@ AFRAME.registerComponent("avatar-audio-source", {
...
+    if (!stream) return;
+
+    getOwnerId(this.el).then(async ownerId => {
+      if (ownerId === peerId && kind === "audio") {
+        // The audio stream for this peer has been updated
+        const newStream = await NAF.connection.adapter.getMediaStream(peerId, "audio").catch(e => {
+          console.error(INFO_INIT_FAILED, `Error getting media stream for ${peerId}`, e);
+        });
+
+        if (newStream) {
```
Comment (netpro2k) [inline src/components/avatar-audio-source.js:117]: Nice this feels much simpler than before. Important that we keep this as simple as possible since I could easily see this being a point of failure for "audio issues"

File: src/components/media-views.js [hunk 10]
```diff
@@ -694,6 +694,20 @@ AFRAME.registerComponent("media-video", {
       if (url.startsWith("hubs://")) {
         const streamClientId = url.substring(7).split("/")[1]; // /clients/<client id>/video is only URL for now
         const stream = await NAF.connection.adapter.getMediaStream(streamClientId, "video");
+        // We subscribe to video stream notifications for this peer to update the video element
+        // This could happen in case there is an ICE failure that requires a transport recreation.
+        this._onStreamUpdated = async (peerId, kind) => {
```
Comment (netpro2k) [inline src/components/media-views.js:699]: Don't think this will actually happen in practice since I don't think we ever change the `src` for a video running from a MediaStream, but it might be worth checking if `this._onStreamUpdated` is non null here and removing the listener before  registering the new one.

File: src/naf-dialog-adapter.js [hunk 11]
```diff
@@ -192,161 +188,122 @@ export default class DialogAdapter {
...
+    }
+
+    try {
+      if (!this._sendTransport?._closed) {
+        // Renegotiate ICE in case ICE state is "failed".
+        // We also try to renegotiate in case we got stuck in a new state after an ICE failure.
+        if (
+          force ||
+          this._sendTransport.connectionState === "failed" ||
+          (this._sendTransport.connectionState === "new" && this._lastSendConnectionState === "failed")
```
Comment (netpro2k) [inline src/naf-dialog-adapter.js:224]: Not too sure what state this is checking for

File: src/naf-dialog-adapter.js [hunk 12]
```diff
@@ -192,161 +188,122 @@ export default class DialogAdapter {
...
+  async recreateSendTransport(iceServers) {
+    this.emitRTCEvent("log", "RTC", () => `Recreating send transport ICE`);
+    await this.closeSendTransport();
+    await this.createSendTransport(iceServers);
+  }
 
   /**
    * Restart ICE in the underlying send peerconnection.
    */
   async restartSendICE(force = true) {
```
Comment (netpro2k) [inline src/naf-dialog-adapter.js:211]: Feels slightly weird to have this force argument. I think its in part a holdover from when this was being directly used as part of the watchdogs. Instead I would argue this function should always restart ice if called (or recreate the transport if its been closed I guess, still want to understand what that case is). Then we can handle any other logic about when we actually want to restart ice outside of this (which I think we are already doing in `checkXIceStatus`.

File: src/naf-dialog-adapter.js [hunk 13]
```diff
@@ -519,9 +442,55 @@ export default class DialogAdapter {
...
+
+      if (pendingMediaRequests.video) {
+        pendingMediaRequests.video.resolve(null);
+      }
+
+      this._pendingMediaRequests.delete(peerId);
+    }
+
+    // Resolve initial audio resolver since this person left.
+    const initialAudioResolver = this._initialAudioConsumerResolvers.get(peerId);
```
Comment (netpro2k) [inline src/naf-dialog-adapter.js:475]: Not part of this PR but the whole `_initialAudioConsumerResolvers` thing is something I want to look into soon. This stuff still feels a bit fishy to be and would like to avoid it if we can. I would bet there are cases where we end up stalling on this for whatever reason. Its unclear why we need to wait for this.

File: src/naf-dialog-adapter.js [hunk 14]
```diff
@@ -613,141 +582,197 @@ export default class DialogAdapter {
     // Not implemented
   }
 
-  async _joinRoom() {
-    debug("_joinRoom()");
-
-    try {
-      this._mediasoupDevice = new mediasoupClient.Device({});
+  async createSendTransport(iceServers) {
```
Comment (netpro2k) [inline src/naf-dialog-adapter.js:585]: Didn't comb through the changes here very closely since I think this is just moving code around, let me know if that is not the case.

File: src/naf-dialog-adapter.js [hunk 15]
```diff
@@ -613,141 +582,197 @@ export default class DialogAdapter {
...
-      this._recvTransport = this._mediasoupDevice.createRecvTransport({
-        id: recvTransportInfo.id,
-        iceParameters: recvTransportInfo.iceParameters,
-        iceCandidates: recvTransportInfo.iceCandidates,
-        dtlsParameters: recvTransportInfo.dtlsParameters,
-        sctpParameters: recvTransportInfo.sctpParameters,
-        iceServers,
-        iceTransportPolicy: this._iceTransportPolicy
-      });
+  async createRecvTransport(iceServers) {
```
Comment (netpro2k) [inline src/naf-dialog-adapter.js:689]: Don't think we should do it in this PR but just noting for future cleanup. Seems like we could DRY up creating the send/receive transports. A lot of the code in here is the same for both.

**Review by netpro2k — COMMENTED** [review]
Pretty happy with where we landed on this, nice work! I think the refactoring of naf-dialog-adapter is a nice step forward in starting to make it easier to reason about, and I think removing the watchdogs and handing stream recreation should hopefully lead to more reliability. I made some suggestions on future cleanup work we can do but I think most of it can be done in future PRs and we don't need to hold this back. I am still a bit confused on some of the stuff in `restartSendICE` and would like to discuss that, but other than that lgtm!

File: src/naf-dialog-adapter.js (hunk 11 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:224]: That's just a state I've seen a couple of time happening. I can just remove it and keep it simple until we have more insights about WebebRTC errors.

File: src/naf-dialog-adapter.js (hunk 12 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:211]: Yep, that argument doesn't make much sense now, I'm going to remove it too.

File: src/components/media-views.js (hunk 10 above)
Comment (keianhzo) [reply src/components/media-views.js:699]: Updated

File: src/naf-dialog-adapter.js (hunk 13 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:475]: I guess that covers the case where that peer disconnected before it's audio consumer resolved which is something that we do upon connect to wait until we receive all the other peer's consumers. I agree that can cause some issues. I'm not sure why we do that though, is it really that important that we wait until we can hear all the other peers to connect?

**Comment by keianhzo** [comment]
@netpro2k Updated with some of the change requests above. If you are ok, I'll move forward with landing this.

File: src/components/media-views.js [hunk 16]
```diff
@@ -694,6 +694,23 @@ AFRAME.registerComponent("media-video", {
...
+          NAF.connection.adapter.off("stream_updated", this._onStreamUpdated);
+        }
+        this._onStreamUpdated = async (peerId, kind) => {
+          if (peerId === streamClientId && kind === "video") {
+            // The video stream for this peer has been updated
+            const stream = await NAF.connection.adapter.getMediaStream(peerId, "video").catch(e => {
+              console.error(`Error getting video stream for ${peerId}`, e);
+            });
+            if (stream) {
+              videoEl.srcObject = new MediaStream(stream);
```
Comment (gfodor) [inline src/components/media-views.js:709]: Hi! I discovered this doesn't work unless you also call `videoEl.play()`, because if the track dies the video is stopped.

Comment (keianhzo) [reply src/components/media-views.js:709]: Does this happen in all browsers? Can you provide STRs? I can can't really reproduce it with current master in FF/Chrome.

The `stream_update` event in **media-view.js** and **avatar-audio-source.js** should only be as a consequence of a _receive transport recreation_, otherwise both components are not  yet created so the `stream_update` event is not triggered on them. Maybe we are are missing some case where it's being triggered? I  haven't found any but having STRs would really help.

A _receive transport recreation_ should only happen in case where the transport has been closed but signaling is still opened. The transport disconnected/failed states are handled by the WebRTC stack and we should recover without any consumer recreation or transport updates whatsoever (so no `stream_update` events). 

I couldn't also reproduce the issue mentioned above. How are you triggering it?
